pkg/predicate/predicate.go

```diff
@@ -0,0 +1,193 @@
+		},
+	}
+}
+
+// LastOperationNotSuccessful is a predicate for unsuccessful last operations for creation events.
+func LastOperationNotSuccessful() predicate.Predicate {
+	operationNotSucceeded := func(obj runtime.Object) bool {
+		etcd, ok := obj.(*druidv1alpha1.Etcd)
+		if !ok {
+			return false
+		}
+		if etcd.Status.LastError != nil {
+			return true
+		}
+		return false
+	}
+
+	return predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			return operationNotSucceeded(event.Object)
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			return operationNotSucceeded(event.ObjectNew)
+		},
+		GenericFunc: func(event event.GenericEvent) bool {
+			return operationNotSucceeded(event.Object)
+		},
+		DeleteFunc: func(event event.DeleteEvent) bool {
+			return operationNotSucceeded(event.Object)
+		},
+	}
+}
+
+// GenerationChangedPredicate implements a default update predicate function on Generation change.
+//
+// This predicate will skip update events that have no change in the object's metadata.generation field.
+// This predicate will skip create events that have the object's metadata.generation does not match status.observedGeneration field.
+// The metadata.generation field of an object is incremented by the API server when writes are made to the spec field of an object.
+// This allows a controller to ignore update events where the spec is unchanged, and only the metadata and/or status fields are changed.
+//
+// For CustomResource objects the Generation is only incremented when the status subresource is enabled.
+//
+// Caveats:
+//
+// * The assumption that the Generation is incremented only on writing to the spec does not hold for all APIs.
+// E.g For Deployment objects the Generation is also incremented on writes to the metadata.annotations field.
+// For object types other than CustomResources be sure to verify which fields will trigger a Generation increment when they are written to.
+//
+// * With this predicate, any update events with writes only to the status field will not be reconciled.
+// So in the event that the status block is overwritten or wiped by someone else the controller will not self-correct to restore the correct status.
+type GenerationChangedPredicate struct {
+	predicate.Funcs
+}
+
+// Update implements default UpdateEvent filter for validating generation change
+func (GenerationChangedPredicate) Update(e event.UpdateEvent) bool {
+	if e.MetaOld == nil {
+		logger.Error("update event has no old metadata", "event", e)
+		return false
+	}
+	if e.ObjectOld == nil {
+		logger.Error("update event has no old runtime object to update", "event", e)
+		return false
+	}
+	if e.ObjectNew == nil {
+		logger.Error("update event has no new runtime object for update", "event", e)
+		return false
+	}
+	if e.MetaNew == nil {
+		logger.Error("update event has no new metadata", "event", e)
+		return false
+	}
+	if e.MetaNew.GetGeneration() == e.MetaOld.GetGeneration() {
+		return false
+	}
+	return true
```

Comment:
Shouldn't we add the `etcd.Status.ObservedGeneration != nil && *etcd.Status.ObservedGeneration == etcd.Generation` check here as well, just to be safe?

Reply:
In case of an update, the oldMeta and newMeta.Generation would be different. So we needn't compare with the status. In case of create, we dont have an oldMeta to compare against.  Create is also triggered when etcd-druid startup. So I am using the observedGeneration.

Reply:
I know. Additional safety check won't hurt, no? What if there is a but in upstream libraries? We know client-go watch timeout implementation sucked.